Prerequisite Changes For USDShade Support

danieldresser-ie opened the pull request:
The SceneWriter destructor change is extremely straightforward, and I'm quite confident of it.

Storing values on shader parameters with inputs is also pretty straightforward, but there are more potential knock-on effects ... everything seems to be working fine though.

danieldresser-ie commented:
I got distracted by thinking about how Splines should work, and didn't do quite as much testing of this as I should have, and got a test failure on CI.  The 3rd commit is necessary in order for `InteractiveArnoldRenderTest.testQuadLightTextureEdits`.  The fact that this change does cause issues, and that only 1 test caught it ( which was not intended to be a test for Arnold parameter handling ), is making me feel much more nervous about this change.

I haven't fully assessed what's going on, but I think what's happening is that if we've declared an RGBA value on the shader, Arnold then refuses to connect it to an RGB source, whereas if we just declare a value, Arnold will auto-convert?  Or maybe it's not about the mismatched types, and all Arnold shader connections were broken without this fix?  But I would have hoped more than one test would have picked that up.  Anyway, more investigation tomorrow.

johnhaddon commented on src/GafferScene/Shader.cpp:
```diff
@@ -455,11 +455,20 @@ class Shader::NetworkBuilder
 			}
 			else
 			{
+				// Even though we have an input connection, still store the value of the parameter on
+				// this node.  This allows us to keep track of the type of the connection, and is more
+				// aligned with how other shader network storage formats, like USD, work.
+				if( IECore::DataPtr value = static_cast<const Shader *>( parameter->node() )->parameterValue( parameter ) )
+				{
+					shader->parameters()[parameterName] = value;
+				}
```
This change needs to be reflected in `hashParameter()` too. It feels like maybe this could also be simplified a fair bit now that both sides of the conditional are generating a parameter value _and_ processing connections in some way. Maybe we can refactor so that we don't need the conditional at all?

danieldresser-ie replied:
Cleaned up a bit and fixed hashParameter in 5728fe7

I've also done a bit more of a refactor in 3cbf483, that feels a little simpler to me in general?

danieldresser-ie replied:
The test failure unfortunately suggests an actual problem.  In our conversation this morning, if I recall it correctly, you said that when writing out a value on a parameter, we should always write out the local plug value.  If it's driven by an output, we make a real conneciton.  If that plug value is driven by something that isn't an output plug, then Gaffer's evaluation should provide an appropriate value for it, and this makes sure the value is actually of the right type ( as opposed to the value of the source plug, which could be a different type ).

Based on this reasoning, I did my refactor, and was feeling pretty happy with it in terms of cleanliness - I was able to have the parameter value always driven by the local plug value instead of being in the conditional, like you wanted.  But the test cases point out that I missed an aspect here:  disabled shaders are supposed to pass through a value based on correspondingInput, and the compute on a shader does not do this passthrough.

My first thought is that maybe the compute on a shader should return a passthrough value when its disabled, but metadata isn't supposed to affect computes, and that could have other weird effects.

Maybe I should just go back to the conditional that sometimes uses effectiveParameter that I started with ... though your point about it being wrong that this could pick up the wrong plug type stands.

johnhaddon commented on src/IECoreArnold/ShaderNetworkAlgo.cpp:
```diff
@@ -198,65 +198,73 @@ AtNode *convertWalk( const ShaderNetwork::Parameter &outputParameter, const IECo
-		if( !sourceNode )
+		if( connected.count( namedParameter.first ) )
 		{
+			// Don't set values on parameters that are connected
```
This rearrangement of the logic is a bit misleading I think - we're still setting values on parameters that have connections at the component (`color.r`) level. This is completely necessary, and was more obviously intentional with the previous logic. If we can make it work, I definitely prefer the approach of just transferring all the values we have, and all the connections.

danieldresser-ie replied:
I've confirmed that the issue seems to be particular to quad_light.color.  We could do a specific fix, but your proposed solution of calling reset when we connect an input is easy enough that maybe there's no harm in doing it unconditionally.  I've done that here: 158da5e

johnhaddon reviewed (commented):
> I haven't fully assessed what's going on, but I think what's happening is that if we've declared an RGBA value on the shader, Arnold then refuses to connect it to an RGB source, whereas if we just declare a value, Arnold will auto-convert? Or maybe it's not about the mismatched types, and all Arnold shader connections were broken without this fix? But I would have hoped more than one test would have picked that up. Anyway, more investigation tomorrow.

Arnold seems to have some special-case handling of quad light colours that we've had to work around before - see https://github.com/GafferHQ/gaffer/commit/5f1316bb489a7a92ba4bf9c94ae90c62b45dab36. I wonder if that is also the cause of the problem here, and could explain why it is limited to this one test case?

johnhaddon reviewed (changes requested):
> Maybe I should just go back to the conditional that sometimes uses effectiveParameter that I started with ... though your point about it being wrong that this could pick up the wrong plug type stands.

I had a quick look at this myself and without making much more drastic changes I can't see anything to improve on that. So yes, let's go back to the original approach. Sorry to have led you on a wild goose chase.

Would be good to include a test cases for the new values-when-connected behaviour though, so we have more of a safety net if we do want to refactor this in future...

johnhaddon commented:
In better news, I've confirmed that this PR fixes the crashes I was seeing when I was using the Cortex PR on its own.

danieldresser-ie commented on src/IECoreArnold/ShaderNetworkAlgo.cpp:
```diff
@@ -254,6 +254,12 @@ AtNode *convertWalk( const ShaderNetwork::Parameter &outputParameter, const IECo
+		// on "quad_light" though, where the connection is ignored unless parameter is reset
+		// first.  We could do this reset selectively just for this node and parameter, but
+		// there's probably nothing wrong with doing it unconditionally
+		AiNodeResetParameter( node, parameterName.c_str() );
```
Addressed and squashed.

danieldresser-ie commented:
I've reverted to the previous code for setting parameters with connections ( and fixed the hashing, which wasn't correct originally ).

On further reflection, I think the new way I was doing things could actually be the best way, it just requires adding the passthrough in the compute for disabled shaders.  The hesitation I expressed was around correspondingInput depending on metadata - but I was just confused, that's shader metadata not Gaffer metadata, so it's OK to use it in a compute - all of this is happening in a compute anyway.  So maybe that would be a good cleanup, but it would be more changes than are strictly required to get the USDShade stuff working.

danieldresser-ie commented:
I caught one more issue:  the same issue with quad_light also affects skydome_light.  Makes sense, but I hope there aren't any more gotchas that will catch us in IPR.

The fix is straightforward so I've applied it in place to the existing PR.
